scripts/update_appup.escript

```diff
@@ -1,75 +1,350 @@
+        OldActions ++
+        [{delete_module, M} || M <- Deleted].
+
+%% @doc Process the existing actions to exclude modules that are
+%% already handled
+process_old_action({purge, Modules}) ->
+    Modules;
+process_old_action({delete_module, Module}) ->
+    [Module];
+process_old_action(LoadModule) when is_tuple(LoadModule) andalso
+                                    element(1, LoadModule) =:= load_module ->
+    element(2, LoadModule);
+process_old_action(_) ->
+    [].
+
+ensure_pred_versions(PredVersion, Versions) ->
+    {Maj, Min, Patch} = parse_semver(PredVersion),
+    PredVersions = [semver(Maj, Min, P) || P <- lists:seq(0, Patch)],
+    lists:foldl(fun ensure_version/2, Versions, PredVersions).
+
+ensure_version(Version, Versions) ->
+    case lists:keyfind(Version, 1, Versions) of
+        false ->
+            [{Version, []}|Versions];
+        _ ->
+            Versions
+    end.
 
 read_appup(File) ->
-    {ok, Bin0} = file:read_file(File),
-    %% Hack:
-    Bin1 = re:replace(Bin0, "VSN", "\"VSN\""),
-    TmpFile = filename:join("/tmp", filename:basename(File)),
-    ok = file:write_file(TmpFile, Bin1),
-    {ok, [Terms]} = file:consult(TmpFile),
-    Terms.
+    case file:script(File, [{'VSN', "VSN"}]) of
+        {ok, Terms} ->
+            Terms;
+        Error ->
+            fail("Failed to parse appup file ~s: ~p", [File, Error])
+    end.
+
+diff_releases(Curr, Old) ->
+    Fun = fun(App, Modules, Acc) ->
+                  OldModules = maps:get(App, Old, #{}),
+                  Acc#{App => diff_app_modules(Modules, OldModules)}
+          end,
+    maps:fold(Fun, #{}, Curr).
+
+diff_app_modules(Modules, OldModules) ->
+    {New, Changed} =
+        maps:fold( fun(Mod, MD5, {New, Changed}) ->
+                           case OldModules of
+                               #{Mod := OldMD5} when MD5 =:= OldMD5 ->
+                                   {New, Changed};
+                               #{Mod := _} ->
+                                   {New, [Mod|Changed]};
+                               _ -> {[Mod|New], Changed}
+                           end
+                   end
+                 , {[], []}
+                 , Modules
+                 ),
+    Deleted = maps:keys(maps:without(maps:keys(Modules), OldModules)),
+    {New, Changed, Deleted}.
+
+find_beams(Dir) ->
+    [filename:join(Dir, I) || I <- filelib:wildcard("**/ebin/*.beam", Dir)].
+
+prepare(Baseline, Options = #{make_command := MakeCommand, beams_dir := BeamDir}) ->
+    log("~n===================================~n"
+        "Baseline: ~s"
+        "~n===================================~n", [Baseline]),
+    log("Building the current version...~n"),
+    bash(MakeCommand),
+    log("Downloading and building the previous release...~n"),
+    {ok, PredRootDir} = build_pred_release(Baseline, Options),
+    {BeamDir, filename:join(PredRootDir, BeamDir)}.
+
+build_pred_release(Baseline, #{clone_url := Repo, make_command := MakeCommand}) ->
```

Comment:
I wonder if we can download a zip package.

Reply:
Yes, definitely. I just want to move it to a separate PR, if possible.